src/shared/mount-util.c

```diff
@@ -797,6 +797,98 @@ int mount_option_mangle(
         return 0;
 }
 
+static int mount_in_namespace_new(
+                const char *chased_src_path,
+                int chased_src_fd,
+                struct stat *chased_src_st,
+                const char *dest,
+                int pidns_fd,
+                int mntns_fd,
+                int root_fd,
+                bool read_only,
+                bool make_file_or_directory) {
+
+        _cleanup_close_pair_ int errno_pipe_fd[2] = PIPE_EBADF;
+        _cleanup_close_ int new_mount_fd = -EBADF;
+        pid_t child;
+        int r;
+
+        assert(chased_src_path);
+        assert(chased_src_fd >= 0);
+        assert(chased_src_st);
+        assert(dest);
+        assert(pidns_fd >= 0);
+        assert(mntns_fd >= 0);
+        assert(root_fd >= 0);
+
+        new_mount_fd = open_tree(
+                        chased_src_fd,
+                        "",
+                        OPEN_TREE_CLONE|OPEN_TREE_CLOEXEC|AT_SYMLINK_NOFOLLOW|AT_EMPTY_PATH);
+        if (new_mount_fd < 0)
+                return log_debug_errno(
+                                errno,
+                                "Failed to open mount point \"%s\": %m",
+                                chased_src_path);
+
+        if (read_only) {
+                if (mount_setattr(new_mount_fd, "", AT_EMPTY_PATH,
+                                &(struct mount_attr) {
+                                        .attr_set = MOUNT_ATTR_RDONLY,
+                                }, MOUNT_ATTR_SIZE_VER0) < 0)
+                        return log_debug_errno(
+                                        errno,
+                                        "Failed to set mount flags for \"%s\": %m",
+                                        chased_src_path);
+        }
+
+        if (pipe2(errno_pipe_fd, O_CLOEXEC|O_NONBLOCK) < 0)
+                return log_debug_errno(errno, "Failed to create pipe: %m");
+
+        r = namespace_fork("(sd-bindmnt)",
+                           "(sd-bindmnt-inner)",
+                           /* except_fds= */ NULL,
+                           /* n_except_fds= */ 0,
+                           FORK_RESET_SIGNALS|FORK_DEATHSIG,
+                           pidns_fd,
+                           mntns_fd,
+                           /* netns_fd= */ -1,
+                           /* userns_fd= */ -1,
```

Comment:
hmm, shouldn't we also enter new userns? (just asking, i really don't know?)

Reply:
Doesn't seem to be needed, at least with the tests that we have are working fine for both images and bind mounts - @brauner ?

Reply:
As long as the caller is privileged wrt to the owning user namespace of the target mount namespace it should work.

Reply:
yes in this case it will always be, as it creates the namespace in the first place, so we should be ok, thanks